decisions/DR004-Core-language-support.md

```diff
@@ -0,0 +1,476 @@
+    static Handle create(Args &&... args)
+    {
+        try
+        {
+            auto pp = new Ptr<Class>;
+            *pp = std::make_shared<Class>(std::forward<Args &&>(args)...);
+            return reinterpret_cast<Handle>(pp);
+        }
+        catch (...)
+        {
+            return nullptr;
+        }
+    }
+
+    template <class Class, class Handle>
+    static void release(Handle handle)
+    {
+        auto pp = reinterpret_cast<Ptr<Class> *>(handle);
+        delete pp;
+    }
+
+    template <class Class, class Handle>
+    static Ptr<Class> & to_cpp(Handle handle)
+    {
+        return *reinterpret_cast<Ptr<Class> *>(handle);
+    }
+
+    template <class Handle, class Class>
+    static Handle borrow_to_c(Ptr<Class> const & p)  // Note: must `release()` when done.
+    {
+        auto pp = new Ptr<Class>{p};
+        return reinterpret_cast<Handle>(pp);
+    }
+};  // namespace CHandle
+}  // namespace openassetio
+
+////////////////////////////////////////////////////////////////////////
+// StringMap
+
+namespace openassetio
+{
+inline namespace OpenAssetIO_VERSION
+{
+using StringMap = std::unordered_map<std::string, std::string>;
+}
+}  // namespace openassetio
+
+#define ns(thing) OpenAssetIO_C_NAMESPACE(thing)
+
+extern "C"
+{
+    // Opaque C handle to a pointer to a shared_ptr to a StringMap
+    typedef struct ns(StringMapOpaque) * ns(StringMapHandle);
+
+    // Create
+    ns(StringMapHandle) ns(StringMap_create)()
+    {
+        using openassetio::CHandle;
+        using openassetio::StringMap;
+        return CHandle::create<StringMap, ns(StringMapHandle)>();
+    }
+
+    // Release
+    void ns(StringMap_release)(ns(StringMapHandle) handle)
+    {
+        using openassetio::CHandle;
+        using openassetio::StringMap;
+        CHandle::release<StringMap>(handle);
+    }
+
+    // Setter
+    void ns(StringMap_set)(ns(StringMapHandle) handle, const char * key, const char * val)
+    {
+        using openassetio::CHandle;
+        using openassetio::StringMap;
+        auto & ptrStringMap = CHandle::to_cpp<StringMap>(handle);
+        ptrStringMap->insert({key, val});
+    }
+}
+
+////////////////////////////////////////////////////////////////////////
+// Dummy Manager declaration
+
+#undef ns
+#define ns(thing) OpenAssetIO_C_NAMESPACE(hostAPI_##thing)
+
+namespace openassetio
+{
+inline namespace OpenAssetIO_VERSION
+{
+namespace hostAPI
+{
+struct Manager
+{
+    void updateTerminology(StringMap & terminology) const;
+    std::string displayName() const;
+};
+}  // namespace hostAPI
+}  // namespace OpenAssetIO_VERSION
+}  // namespace openassetio
+extern "C"
+{
+    typedef struct ns(ManagerOpaque) * ns(ManagerHandle);
+}
+
+////////////////////////////////////////////////////////////////////////
+// Default terminology
+
+#undef ns
+#define ns(thing) OpenAssetIO_C_NAMESPACE(terminology_##thing)
+
+extern "C"
+{
+    [[maybe_unused]] static const char * ns(kTerm_Asset) = "asset";
```

Comment:
years of fighting MSVC and string deduplication, and how to decorate extern statics for __declspec(dllexport) make me strongly prefer strings not being inlined like this.

I'd go so far as to say that both char* and std::string are more or less anti-patterns... Moving forward, I'm using the pattern of library specific string classes, with c_str and std::string adaptors.

the reason is that we generally want to impose constraints on allowable strings, such as notions of being well formed. all that goes out the window with the standard types, and all the many ways one can introduce a malformed string into the system become an ongoing maintenance concern.

A typical example is a string to represent a URL. If it's a string, it has to be validated and sanitized at every point of use. If it's a specialized value, it's sanitized/validated on construction, and can subsequently make a promise of being well formed.

Reply:
+1 to your own string type. It doesn't actually have to support may operations if it can convert to/from std::string and char* and it gives you compiler independence for movability, alignment and size.